src/ray/gcs/pubsub/gcs_pub_sub.h

```diff
@@ -165,5 +165,69 @@ class RAY_EXPORT PythonGcsPublisher {
   int gcs_port_;
 };
 
+// This client is only supposed to be used from Cython / Python
+class RAY_EXPORT PythonGcsSubscriber {
+ public:
+  explicit PythonGcsSubscriber(const std::string &gcs_address,
+                               rpc::ChannelType channel_type,
+                               const std::string &subscriber_id,
+                               const std::string &worker_id);
+
+  /// Connect to the subscriber service of the GCS.
+  /// This function must be called before calling other functions.
+  ///
+  /// \return Status
+  Status Connect();
+
+  /// Register a subscription for the subscriber's channel type.
+  ///
+  /// Before the registration, published messages in the channel
+  /// will not be saved for the subscriber.
+  Status Subscribe();
+
+  /// Polls for new error message.
+  /// Both key_id and data are out parameters.
+  Status PollError(std::string *key_id, int64_t timeout_ms, rpc::ErrorTableData *data);
+
+  /// Polls for new log messages.
+  Status PollLogs(std::string *key_id, int64_t timeout_ms, rpc::LogBatch *data);
+
+  /// Polls for new function key messages.
+  Status PollFunctionKey(std::string *key_id,
+                         int64_t timeout_ms,
+                         rpc::PythonFunction *data);
+
+  /// Polls for actor messages.
+  Status PollActor(std::string *key_id, int64_t timeout_ms, rpc::ActorTableData *data);
+
+  /// Closes the subscriber and its active subscription.
+  Status Close();
+
+  int64_t last_batch_size();
+
+ private:
+  Status DoPoll(int64_t timeout_ms, rpc::PubMessage *message);
+
+  mutable absl::Mutex mu_;
+
+  std::unique_ptr<rpc::InternalPubSubGcsService::Stub> pubsub_stub_;
+  std::shared_ptr<grpc::Channel> channel_;
+  std::string gcs_address_;
+  int gcs_port_;
+  rpc::ChannelType channel_type_;
+  std::string subscriber_id_;
+  std::string publisher_id_;
+  std::string worker_id_;
+  int64_t max_processed_sequence_id_ GUARDED_BY(mu_);
+  int64_t last_batch_size_ GUARDED_BY(mu_);
+  std::deque<rpc::PubMessage> queue_ GUARDED_BY(mu_);
+  bool closed_ GUARDED_BY(mu_);
+  std::shared_ptr<grpc::ClientContext> current_polling_context_ GUARDED_BY(mu_);
```

Comment:
Wait why is this a `shared_ptr`? Specifically so that we can make a readable shared_ptr copy under lock? But that's a shallow copy so then when we unlock and call `get()` it's still going to get from the same address that is now not guarded by anything so it'll still be a data race. Maybe if we want to do it this way we should just copy the value and get rid of shared_ptr altogether?

Reply:
The shared_ptr here is purely because `grpc::ClientContext` cannot be copied. You were right that I originally modified the value the shared pointer was pointing to, which could potentially produce a data race. I changed it now to not overwrite the existing shared pointer and instead produce a new one, so the object that the private shared_ptr `context` is pointing to can never be overwritten while the RPC is active.

Reply:
Nice, yeah

(Side note that folly has hazard pointers and RCU to deal with this kind of thing and in a lock-free way)